### PR TITLE
[draft] Update to govuk frontend 5.7.1

### PR DESCRIPTION
**Pull request by RogerHowellDfE**
## Intent

Currently this targets govuk-frontend v5.2.0 (released 21 Feb 2024).

This draft PR is work towards supporting govuk-frontend v5.7.1 (released 11 Oct 2024).

Out of scope of this PR are new tag helpers such as the password component (new in v5.3.0 #286 ). Ditto for older components such as the header (#18 ) or footer (#17 ) tag helpers.

## Changes

- [x] Remove redundant `role` attributes, per changes in govuk-frontend v5.3.0
- [ ] Update HTML structure of character count component, per changes in govuk-frontend v5.3.0
- [x] Update test to check for presence of `svg` crest logo (previously `png`, per changes in v5.7.0

These changes are also flagged as changes per test suites, but seemingly no explicit mention in the govuk-frontend changelog.

- [x] Update error summary to not output an empty `<ul>`, per conformance test expectations
- [x] Handle `autocapitalize` on text input components, per conformance test expectations
- [ ] TBC: conformance test failure on the `with customised input wrapper` case

Note: I did begin to do an incremental upgrade but all the headaches seem to be introduced in v5.3.0 (character count component) which makes this idea not so useful 🤷 .

### CODE REVIEW

**Comment by RogerHowellDfE** [comment]
There are several minor changes in more recent versions, but the character component changes (introduced in v5.3.0) are proving to be a headache.

Specifically, this seems to be the first case where the `govuk-form-group` and the top-most / outermost element is shared / used as both the container and the "named" element.

This, unfortunately, then seems to break many of the assumptions/logic in the helper methods/functionality which assumes they are at separate levels of nesting 
- e.g., methods on `FormGroupTagHelperBase` don't necessarily have the `*Context`s available
  - the `*Context` is where data such as the max counts etc. are stored, and 
  - where `*Context` is provided, it doesn't seem to have access to the parent element
- e.g., test helpers such as `GenerateFormGroup` don't seem to allow easy editing of the root element?).

Being mostly new to the internals, there is a decent chance I am missing something obvious however! 🥲 
- Similarly, I might be hamstringing myself by not wanting to make too substantial a change to method signatures etc.
- ... feel free to brutally wave away the proposed changes and signpost to a better approach!

File: tests/GovUk.FrontEnd.AspNetCore.TestCommon/AssertEx.cs [hunk 1]
```diff
@@ -78,7 +78,10 @@ public static void HtmlEqual(
                 DiffConverter.Append(diff, sb);
             }
 
-            throw new XunitException(sb.ToString());
+            throw new XunitException("---\nActual: \n---\n" + actual.Replace(">", ">\n").Replace("<", "\n<").Replace(">\n\n<", ">\n<").Trim() + "\n\n" +
+                                     "---\nExected: \n---\n" + expected + "\n\n" +
+                                     "---\n" +
+                                     "" + sb.ToString());
```
Comment (RogerHowellDfE) [inline tests/GovUk.FrontEnd.AspNetCore.TestCommon/AssertEx.cs:84]: The HTML diff is very useful/required to flag where the differences are, but it only gets so far if not able to visualise the broader context i.e. the syntax tree / html string it is referring to (a-la "stop at the third red house on the left").

I have found it useful to be able to quickly reference expected/actual HTML, to then try and figure out whether it is the container versus a nested element (or both) whose behaviour is wrong.

It's a bit of a hacky solution and might not be desirable to merge this, however?

File: tests/GovUk.Frontend.AspNetCore.ConformanceTests/ComponentTests.cs [hunk 2]
```diff
@@ -39,16 +55,110 @@ private static void AppendToDescribedBy(ref string describedBy, string value)
...
+            _outputHelper.WriteLine($"Actual HTML written to: {actualFilePath}");
+        }
+        catch (Exception ex)
+        {
+            throw new InvalidOperationException($"Failed to write HTML files to {outputBaseDirectory}.", ex);
+        }
+
+        // Assert that the generated HTML matches the expected HTML
         AssertEx.HtmlEqual(testCaseData.ExpectedHtml, html, excludeDiff);
     }
```
Comment (RogerHowellDfE) [inline tests/GovUk.Frontend.AspNetCore.ConformanceTests/ComponentTests.cs:97]: This is some exploratory code where I was writing files to disk (found it easier to diff, and to try and pinpoint what changed/when):

image


There is potentially some utility in storing these within the repo, rather than a temp directory, perhaps? (or maybe just the `Expected` files as this is what is supplied from upstream for each version?)

That's probably a separate discussion/PR though, and I suspect we'll want to avoid merging this for now?

File: tests/GovUk.Frontend.AspNetCore.IntegrationTests/StartupFilterTests.cs [hunk 3]
```diff
@@ -35,7 +35,7 @@ public async Task HostsCompiledAssets(string fileName)
 
     [Theory]
     [InlineData("images/favicon.ico")]
-    [InlineData("images/govuk-crest.png")]
+    [InlineData("images/govuk-crest.svg")]
```
Comment (RogerHowellDfE) [inline tests/GovUk.Frontend.AspNetCore.IntegrationTests/StartupFilterTests.cs:38]: govuk-frontend v5.7.0:

https://github.com/alphagov/govuk-frontend/pull/5376

Note that the CSS changes etc. are all self-contained in the upstream govuk-frontend lib, no need for any additional changes here, beyond the sanity check done as part of this test.


image

File: tests/GovUk.Frontend.AspNetCore.Tests/TagHelpers/ErrorSummaryTagHelperTests.cs [hunk 4]
```diff
@@ -77,6 +77,52 @@ public async Task ProcessAsync_GeneratesExpectedOutput()
...
+        <h2 class=""govuk-error-summary__title"">Title</h2>
+        <div class=""govuk-error-summary__body"">
+            <p>Description</p>
+        </div>
+    </div>
+</div>";
+
+        AssertEx.HtmlEqual(expectedHtml, output.ToHtmlString());
+    }
+
```
Comment (RogerHowellDfE) [inline tests/GovUk.Frontend.AspNetCore.Tests/TagHelpers/ErrorSummaryTagHelperTests.cs:125]: Extra test case to verify behaviour when there are zero error items.

Seems to be introduced in v5.4.0, where the `<ul>` only gets created if the item count is truthy/non-zero:

https://github.com/alphagov/govuk-frontend/pull/4971/files#diff-43521f7999f31d5b19bb203342ab8c5e5ff212b456e0fc072bca7656575e5f6a

image

File: src/GovUk.Frontend.AspNetCore/TagHelpers/CharacterCountTagHelper.cs [hunk 5]
```diff
@@ -194,33 +194,22 @@ public decimal? Threshold
...
-            null,
-            null);
+            textAreaDescriptionText: null,
+            charactersUnderLimitText: null,
+            charactersAtLimitText: null,
+            charactersOverLimitText: null,
+            wordsUnderLimitText: null,
+            wordsAtLimitText: null,
+            wordsOverLimitText: null);
     }
```
Comment (RogerHowellDfE) [inline src/GovUk.Frontend.AspNetCore/TagHelpers/CharacterCountTagHelper.cs:213]: - New argument of `haveError` to make it available in the generator (I don't recall why it is required, but did lots of back-and-forth adding it/removing it/adding it/removing it etc.)
- Renamed argument `formGroup` -> `content` to recognise the `govuk-form-group` is now the root element and not a child
- Given the long run of `null`s, I've added named arguments as an improvement to prevent accidents with positional arguments being mis-positioned.
  - Should these even be null, as opposed to being populated from somewhere?

File: src/GovUk.Frontend.AspNetCore/TagHelpers/CharacterCountTagHelper.cs [hunk 6]
```diff
@@ -254,8 +252,32 @@ private protected override IHtmlContent GenerateFormGroupContent(
...
+                (textAreaDescriptionText ?? $"You can enter up to %{{count}} {(MaxWords.HasValue ? "words" : "characters")}")
+                .Replace("%{count}", (MaxWords.HasValue ? MaxWords : MaxLength).ToString());
+
+            var hintContent = new HtmlString(HtmlEncoder.Default.Encode(content));
+
+            var attributes = CountMessageAttributes.ToAttributeDictionary();
+            attributes.MergeCssClass("govuk-character-count__message");
+
+            return Generator.GenerateHint(hintId, hintContent, attributes);
+        }
```
Comment (RogerHowellDfE) [inline src/GovUk.Frontend.AspNetCore/TagHelpers/CharacterCountTagHelper.cs:279]: Flagging the to-do here. It's one of the examples where there seems to be a mismatch between data being stored on the `*Context` but isn't available on the `FormGroupTagHelperBase#GenerateFormGroupContent` method / overrides.

Note that tests aren't failing with this dodgy implementation, but maybe this means additional tests are needed?

File: src/GovUk.Frontend.AspNetCore/TagHelpers/CharacterCountTagHelper.cs [hunk 7]
```diff
@@ -254,8 +252,32 @@ private protected override IHtmlContent GenerateFormGroupContent(
...
+        contentBuilder.AppendHtml(countRemainingMessage);
+
         return contentBuilder;
 
+
+        IHtmlContent GenerateCountRemainingMessage()
+        {
+            // TODO / FIXME: temporary hardcoded, need to pass through from somewhere? Unable to access context from here?
+            string textAreaDescriptionText = null;
+            var hasNoLimit = MaxLength is null && MaxWords is null; // TODO: This logic is reused in many places -- centralise it?
```
Comment (RogerHowellDfE) [inline src/GovUk.Frontend.AspNetCore/TagHelpers/CharacterCountTagHelper.cs:265]: Assuming the `*Context` could be made available, is it appropriate to centralise the `hasNoLimit` logic to it? Somewhere else, perhaps?

File: src/GovUk.Frontend.AspNetCore/Views/Shared/_GovUkPageTemplate.cshtml [hunk 8]
```diff
@@ -70,7 +70,7 @@
 
     <div class="govuk-width-container @ViewBag.ContainerClasses">
         @RenderSection("BeforeContent", required: false)
-        <main class="govuk-main-wrapper @ViewBag.MainClasses" id="main-content" role="main" lang="@ViewBag.MainLang">
+        <main class="govuk-main-wrapper @ViewBag.MainClasses" id="main-content" lang="@ViewBag.MainLang">
```
Comment (RogerHowellDfE) [inline src/GovUk.Frontend.AspNetCore/Views/Shared/_GovUkPageTemplate.cshtml:73]: v5.3.0

https://github.com/alphagov/govuk-frontend/pull/4854

image

File: tests/GovUk.Frontend.AspNetCore.Tests/TagHelpers/PaginationTagHelperTests.cs [hunk 9]
```diff
@@ -55,7 +55,7 @@ public async Task ProcessAsync_WithPreviousAndNextOnly_GeneratesExpectedOutput()
 
         // Assert
         var expectedHtml = @"
-<nav aria-label=""Pagination"" class=""govuk-pagination--block govuk-pagination"" role=""navigation"">
+<nav aria-label=""Pagination"" class=""govuk-pagination--block govuk-pagination"">
```
Comment (RogerHowellDfE) [inline tests/GovUk.Frontend.AspNetCore.Tests/TagHelpers/PaginationTagHelperTests.cs:58]: v5.3.0

https://github.com/alphagov/govuk-frontend/pull/4854

image

File: tests/GovUk.Frontend.AspNetCore.Tests/TagHelpers/PaginationTagHelperTests.cs [hunk 10]
```diff
@@ -136,7 +136,7 @@ public async Task ProcessAsync_WithItems_GeneratesExpectedOutput()
         // Assert
         var actual = output.ToHtmlString();
         var expectedHtml = @"
-<nav aria-label=""search"" class=""govuk-pagination"" role=""navigation"">
+<nav aria-label=""search"" class=""govuk-pagination"">
```
Comment (RogerHowellDfE) [inline tests/GovUk.Frontend.AspNetCore.Tests/TagHelpers/PaginationTagHelperTests.cs:139]: v5.3.0

https://github.com/alphagov/govuk-frontend/pull/4854

image

**Review by RogerHowellDfE — COMMENTED** [review]
Lots of comments flagging thoughts / TODOs etc.

Apologies it's not all fully finished and polished!

**Comment by RogerHowellDfE** [comment]
Hmm... 

At the time of creating the PR I had only two failing tests (one for an i8n data attribute on the character count component, and one on the text input component in the `with customised input wrapper` scenario).

I clearly must have had something cached funny on my machine and doing a bunch of branch switching just now has unsettled things. 

It seems like there are quite a few additional conformance tests failing due to boolean conversions. It might just be an intermittent thing on my end (workflow is pending approval so can't yet tell from the PR checks) but I'll explore those tomorrow/another day.

**Comment by gunndabad** [comment]
> the character component changes (introduced in v5.3.0) are proving to be a headache.

That's what's stopped me so far from making this upgrade. There's a sizeable refactor of this library going on (see the `dev` branch); I might bring over some of its parts into `main` just to make this change at least do-able.

File: Directory.Build.props [hunk 11]
```diff
@@ -4,9 +4,17 @@
...
+<!--    <GovUkFrontendVersion>5.3.1</GovUkFrontendVersion>-->
+<!--    <GovUkFrontendVersion>5.4.0</GovUkFrontendVersion>-->
+<!--    <GovUkFrontendVersion>5.4.1</GovUkFrontendVersion>-->
+<!--    <GovUkFrontendVersion>5.5.0</GovUkFrontendVersion>-->
+<!--    <GovUkFrontendVersion>5.6.0</GovUkFrontendVersion>-->
+<!--    <GovUkFrontendVersion>5.7.0</GovUkFrontendVersion>-->
+    <GovUkFrontendVersion>5.7.1</GovUkFrontendVersion>
 
-    <_NpmLibDirectory>$([MSBuild]::NormalizeDirectory('$(RepoRoot)', 'lib'))</_NpmLibDirectory>
+    <_NpmLibDirectory>$([MSBuild]::NormalizeDirectory('$(RepoRoot)', 'lib', 'govuk-frontend-$(GovUkFrontendVersion)'))</_NpmLibDirectory>
```
[the remark this answers is not in the document]
Comment (gunndabad) [reply Directory.Build.props:17]: I like this. We could probably do away with the `version.txt` file with this change too.

**Comment by RogerHowellDfE** [comment]
> > the character component changes (introduced in v5.3.0) are proving to be a headache.
> 
> That's what's stopped me so far from making this upgrade. 

Reassuring to hear it's not a "just me" issue having difficulty with the implementation! 😓 😁 

> There's a sizeable refactor of this library going on (see the `dev` branch); I might bring over some of its parts into `main` just to make this change at least do-able.

Apologies I missed the refactor happening on `dev`! 

Is this something you might reasonably want some help with? I appreciate Some changes just need time to work through and adding extra people complicates things, so I promise to not be offended if you say no / not yet!

> Thanks for PR @RogerHowellDfE. I think it would be good to upgrade to 5.3.0 first then upgrade to each frontend version separately, that way I can create a release for each version along the way.

I'm happy to close this PR in favour of this plan. I've had my fun getting to know the library a bit better, so I'll not shed a tear at its loss.

The changes are either trivial (no need to keep this PR for them), or relate to the character count component (not applicable once the refactor comes in?).

Apologies for the unsolicited noise without discussing first! 😄

**Comment by gunndabad** [comment]
@RogerHowellDfE I'm grateful for the help and there are definitely some bits from this PR I'd like to steal :-)

With hindsight having a completely separate branch that's been around for so long wasn't the right thing to do; I'm working on consolidating things now then changes should be easier to add on. I think an iterative approach to the changes I have planned is probably a better way than the big bang I was attempting previously.
